src/normalization.md

```diff
@@ -0,0 +1,314 @@
+
+Normalization is the process of taking these alias types and replacing them with the underlying type that they are equal to. For example given some type alias `type Foo = u32`, normalizing `Foo` would give `u32`.
+
+The concept of an alias is not unique to *types* and the concept also applies to constants/const generics. However, right now in the compiler we don't really treat const aliases as a "first class concept" so this chapter mostly discusses things in the context of types (even though the concepts transfer just fine).
+
+[tykind_alias]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_type_ir/enum.TyKind.html#variant.Alias
+[aliaskind]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_type_ir/enum.AliasTyKind.html
+
+### Rigid, Ambiguous and Unnormalized Aliases
+
+Aliases can either be "rigid", "ambiguous", or simply unnormalized.
+
+We consider types to be rigid if their "shape" isn't going to change, for example `Box` is rigid as no amount of normalization can turn a `Box` into a `u32`, whereas `<vec::IntoIter<u32> as Iterator>::Item` is not rigid as it can be normalized to `u32`.
+
+Aliases are rigid when we know that we will never be able to normalize them further. A concrete example of a *rigid* alias would be `<T as Iterator>::Item` in an environment where there is no `T: Iterator<Item = ...>` bound, only a `T: Iterator` bound:
+```rust
+fn foo<T: Iterator>() {
+    // This alias is *rigid*
+    let _: <T as Iterator>::Item;
+}
+
+fn bar<T: Iterator<Item = u32>>() {
+    // This alias is *not* rigid as it can be normalized to `u32`
+    let _: <T as Iterator>::Item;
+}
+```
+
+This definition also implies that illformed aliases can be considered rigid. For example the type `<u32 as Iterator>::Item` is a rigid alias as it can never be normalized as there is no `u32: Iterator` impl (ignoring `feature(trivial_bounds)`).
+
+When an alias can't yet be normalized but may eventually wind up normalizable we consider it to be an "ambiguous" alias. This can occur when an alias contains inference variables which prevent being able to determine how the trait is implemented:
+```rust
+fn foo<T: Iterator, U: Iterator>() {
+    // This alias is considered to be "ambiguous"
+    let _: <_ as Iterator>::Item;
+}
+```
+
+The reason we call them "ambiguous" aliases is because its *ambiguous* whether this is a rigid alias or not.
+
+The source of the `_: Iterator` trait impl is *ambiguous* (i.e. unknown), it could be some `impl Iterator for u32` or it could be some `T: Iterator` trait bound, we don't know yet. Depending on how `_: Iterator` holds the alias could be an unnormalized alias or it could be a rigid alias; it's *ambiguous* what kind of alias this is.
```

Comment:
```suggestion
The source of the `_: Iterator` trait impl is *ambiguous* (i.e. unknown), it could be some `impl Iterator for u32` or it could be some `T: Iterator` trait bound, we don't know yet. Depending on why `_: Iterator` holds the alias could be an unnormalized alias or it could be a rigid alias; it's *ambiguous* what kind of alias this is.
```